salt/states/x509.py

```diff
@@ -310,45 +330,23 @@ def csr_managed(name,
              - L: Salt Lake City
              - keyUsage: 'critical dataEncipherment'
     '''
-    ret = {'name': name, 'changes': {}, 'result': False, 'comment': ''}
-
-    if os.path.isfile(name):
-        try:
-            current = __salt__['x509.read_csr'](csr=name)
-        except salt.exceptions.SaltInvocationError:
-            current = '{0} is not a valid CSR.'.format(name)
-    else:
-        current = '{0} does not exist.'.format(name)
-
-    new_csr = __salt__['x509.create_csr'](text=True, **kwargs)
-    new = __salt__['x509.read_csr'](csr=new_csr)
-
-    if current == new:
-        ret['result'] = True
-        ret['comment'] = 'The CSR is already in the correct state'
-        return ret
-
-    ret['changes'] = {
-            'old': current,
-            'new': new, }
+    old = __salt__['x509.read_csr'](name)
+    file_args, kwargs = _get_file_args(name, **kwargs)
+    file_args['contents'] = __salt__['x509.create_csr'](text=True, **kwargs)
 
-    if __opts__['test'] is True:
-        ret['result'] = None
-        ret['comment'] = 'The CSR {0} will be updated.'.format(name)
-
-    if os.path.isfile(name) and backup:
-        bkroot = os.path.join(__opts__['cachedir'], 'file_backup')
-        salt.utils.backup_minion(name, bkroot)
-
-    ret['comment'] = __salt__['x509.write_pem'](text=new_csr, path=name, pem_type="CERTIFICATE REQUEST")
-    ret['result'] = True
+    ret = __states__['file.managed'](**file_args)
+    if ret['changes']:
+        new = __salt__['x509.read_csr'](file_args['contents'])
+        if old != new:
+            ret['changes'] = {"Old": old, "New": new}
 
     return ret
 
 
 def certificate_managed(name,
                         days_remaining=90,
-                        backup=False,
+                        managed_private_key=None,
+                        append_certs=[],
```

Comment:
We can't have a mutable type as a default argument. Could we change this to `None` and them modify it inside the function if necessary?